homeassistant/components/modbus/binary_sensor.py

```diff
@@ -99,18 +102,33 @@ def device_class(self) -> Optional[str]:
         """Return the device class of the sensor."""
         return self._device_class
 
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
     def update(self):
         """Update the state of the sensor."""
-        if self._input_type == INPUT_TYPE_COIL:
-            result = self._hub.read_coils(self._slave, self._address, 1)
-        else:
-            result = self._hub.read_discrete_inputs(self._slave, self._address, 1)
+
+        def _set_unavailable():
+            if self._available:
+                _LOGGER.error(
+                    "No response from hub %s, slave %s, address %s",
+                    self._hub.name,
+                    self._slave,
+                    self._address,
+                )
+            self._available = False
+
         try:
+            if self._input_type == INPUT_TYPE_COIL:
+                result = self._hub.read_coils(self._slave, self._address, 1)
+            else:
+                result = self._hub.read_discrete_inputs(self._slave, self._address, 1)
+            if isinstance(result, (ModbusException, ExceptionResponse)):
```

Comment:
huh, exceptions are returned as results and not raised? 🤔

Reply:
This surprised me too. I don't know internal workings of this library. However, it was clear from the error logs that the result object got returned and we were trying to access an attribute. I got errors like `ModbusIOException` has no attribute `register`.

Reply:
Ok, well it's what it is :(

Reply:
I moved the `isinstance(...)` code below the `Exception` clause. I think it looks better now, and the `try...except` block now wraps only calls to `pymodbus`.